File: server/src/services/git.service.ts
import simpleGit, { SimpleGit } from 'simple-git'
import { logger } from '../utils/logger.js'
import { AppError } from '../utils/errors.js'

export class GitError extends AppError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, 500, 'GIT_ERROR')
    if (details) {
      logger.error({ ...details }, message)
    }
  }
}

export interface WorktreeInfo {
  path: string
  branch: string
  isMain: boolean
}

export interface BranchInfo {
  local: string[]
  remote: string[]
  current: string
}

export class GitService {
  private git: SimpleGit

  constructor(repoPath: string) {
    this.git = simpleGit(repoPath)
  }

  async isValidRepository(): Promise<boolean> {
    try {
      await this.git.revparse(['--git-dir'])
      return true
    } catch {
      return false
    }
  }

  async getCurrentBranch(): Promise<string> {
    try {
      const branch = await this.git.revparse(['--abbrev-ref', 'HEAD'])
      return branch.trim()
    } catch (err) {
      throw new GitError('Failed to get current branch', { error: String(err) })
    }
  }

  async listWorktrees(): Promise<WorktreeInfo[]> {
    try {
      const raw = await this.git.raw(['worktree', 'list', '--porcelain'])
      const worktrees: WorktreeInfo[] = []
      let current: Partial<WorktreeInfo> = {}

      for (const line of raw.split('\n')) {
        if (line.startsWith('worktree ')) {
          current.path = line.substring(9)
        } else if (line.startsWith('branch ')) {
          current.branch = line.substring(7).replace('refs/heads/', '')
        } else if (line.startsWith('HEAD ')) {
          // Detached HEAD state - use commit hash
          if (!current.branch) {
            current.branch = line.substring(5).substring(0, 8)
          }
        } else if (line === '') {
          if (current.path) {
            worktrees.push({
              path: current.path,
              branch: current.branch || 'HEAD',
              isMain: worktrees.length === 0,
            })
          }
          current = {}
        }
      }

      // Handle last entry if no trailing newline
      if (current.path) {
        worktrees.push({
          path: current.path,
          branch: current.branch || 'HEAD',
          isMain: worktrees.length === 0,
        })
      }

      return worktrees
    } catch (err) {
      throw new GitError('Failed to list worktrees', { error: String(err) })
    }
  }

  async addWorktree(worktreePath: string, branch: string, createBranch: boolean): Promise<void> {
    try {
      const args = ['worktree', 'add']

      if (createBranch) {
        args.push('-b', branch, worktreePath)
      } else {
        args.push(worktreePath, branch)
      }

      await this.git.raw(args)
      logger.info({ path: worktreePath, branch, createBranch }, 'Worktree created')
    } catch (err) {
      throw new GitError(`Failed to create worktree: ${err}`, {
        path: worktreePath,
        branch,
        error: String(err),
      })
    }
  }

  async removeWorktree(worktreePath: string, force = false): Promise<void> {
    try {
      const args = ['worktree', 'remove']
      if (force) {
        args.push('--force')
      }
      args.push(worktreePath)

      await this.git.raw(args)
      logger.info({ path: worktreePath, force }, 'Worktree removed')
    } catch (err) {
      throw new GitError(`Failed to remove worktree: ${err}`, {
        path: worktreePath,
        error: String(err),
      })
    }
  }

  async checkout(branch: string, createBranch = false): Promise<void> {
    try {
      if (createBranch) {
        await this.git.checkoutLocalBranch(branch)
      } else {
        await this.git.checkout(branch)
      }
      logger.info({ branch, createBranch }, 'Branch checked out')
    } catch (err) {
      throw new GitError(`Failed to checkout branch: ${err}`, {
        branch,
        error: String(err),
      })
    }
  }

  async listBranches(): Promise<BranchInfo> {
    try {
      const local = await this.git.branchLocal()
      const remote = await this.git.branch(['-r'])

      return {
        local: local.all,
        remote: remote.all.map((b) => b.replace('origin/', '')),
        current: local.current,
      }
    } catch (err) {
      throw new GitError('Failed to list branches', { error: String(err) })
    }
  }

  async getStatus(): Promise<{
    isClean: boolean
    ahead: number
    behind: number
    modified: string[]
    staged: string[]
    untracked: string[]
  }> {
    try {
      const status = await this.git.status()
      return {
        isClean: status.isClean(),
        ahead: status.ahead,
        behind: status.behind,
        modified: status.modified,
        staged: status.staged,
        untracked: status.not_added,
      }
    } catch (err) {
      throw new GitError('Failed to get status', { error: String(err) })
    }
  }

  async pull(): Promise<void> {
    try {
      await this.git.pull()
      logger.info('Pull completed')
    } catch (err) {
      throw new GitError('Failed to pull', { error: String(err) })
    }
  }

  async fetch(): Promise<void> {
    try {
      await this.git.fetch()
      logger.info('Fetch completed')
    } catch (err) {
      throw new GitError('Failed to fetch', { error: String(err) })
    }
  }
}

File: src-tauri/src/services/git_service.rs
//! Git service for interacting with git repositories

use git2::{BranchType, Repository, StatusOptions};
use std::path::Path;
use thiserror::Error;

use crate::types::{BranchInfo, GitStatusInfo};

#[derive(Error, Debug)]
pub enum GitError {
    #[error("Git error: {0}")]
    Git(#[from] git2::Error),
    #[error("Not a git repository: {0}")]
    NotARepo(String),
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
}

/// Information about a worktree from git
#[derive(Debug, Clone)]
pub struct WorktreeInfo {
    pub path: String,
    pub branch: String,
    pub is_main: bool,
}

pub struct GitService;

impl GitService {
    /// Check if a path is a valid git repository
    pub fn is_valid_repository(path: &str) -> bool {
        Repository::open(path).is_ok()
    }

    /// Get the current branch name
    pub fn get_current_branch(path: &str) -> Result<String, GitError> {
        let repo = Repository::open(path)?;
        let head = repo.head()?;
        Ok(head.shorthand().unwrap_or("HEAD").to_string())
    }

    /// List all worktrees for a repository
    pub fn list_worktrees(path: &str) -> Result<Vec<WorktreeInfo>, GitError> {
        let repo = Repository::open(path)?;
        let mut worktrees = Vec::new();

        // Main worktree
        let main_path = repo
            .workdir()
            .ok_or_else(|| GitError::NotARepo("No workdir".to_string()))?
            .to_string_lossy()
            .to_string();

        worktrees.push(WorktreeInfo {
            path: main_path.trim_end_matches('/').to_string(),
            branch: Self::get_current_branch(path)?,
            is_main: true,
        });

        // Additional worktrees
        if let Ok(wt_names) = repo.worktrees() {
            for name in wt_names.iter().flatten() {
                if let Ok(wt) = repo.find_worktree(name) {
                    if let Some(wt_path) = wt.path().to_str() {
                        let branch = Self::get_current_branch(wt_path).unwrap_or_default();
                        worktrees.push(WorktreeInfo {
                            path: wt_path.to_string(),
                            branch,
                            is_main: false,
                        });
                    }
                }
            }
        }

        Ok(worktrees)
    }

    /// Add a new worktree
    pub fn add_worktree(
        repo_path: &str,
        worktree_path: &str,
        branch: &str,
        create_branch: bool,
    ) -> Result<WorktreeInfo, GitError> {
        let repo = Repository::open(repo_path)?;

        if create_branch {
            // Create branch from HEAD
            let head = repo.head()?.peel_to_commit()?;
            repo.branch(branch, &head, false)?;
        }

        // Find the branch reference
        let branch_ref = repo.find_branch(branch, BranchType::Local)?;
        let reference = branch_ref.into_reference();

        // Add worktree
        let worktree_name = Path::new(worktree_path)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("worktree");

        repo.worktree(
            worktree_name,
            Path::new(worktree_path),
            Some(git2::WorktreeAddOptions::new().reference(Some(&reference))),
        )?;

        Ok(WorktreeInfo {
            path: worktree_path.to_string(),
            branch: branch.to_string(),
            is_main: false,
        })
    }

    /// Remove a worktree
    pub fn remove_worktree(repo_path: &str, worktree_path: &str) -> Result<(), GitError> {
        let repo = Repository::open(repo_path)?;

        // Find worktree by path
        if let Ok(wt_names) = repo.worktrees() {
            for name in wt_names.iter().flatten() {
                if let Ok(wt) = repo.find_worktree(name) {
                    if wt.path().to_str() == Some(worktree_path) {
                        wt.prune(Some(
                            git2::WorktreePruneOptions::new()
                                .working_tree(true)
                                .valid(true),
                        ))?;

                        // Also remove the directory
                        if Path::new(worktree_path).exists() {
                            std::fs::remove_dir_all(worktree_path)?;
                        }

                        return Ok(());
                    }
                }
            }
        }

        Err(GitError::NotARepo(format!(
            "Worktree not found: {}",
            worktree_path
        )))
    }

    /// Checkout a branch
    pub fn checkout_branch(worktree_path: &str, branch: &str, create: bool) -> Result<(), GitError> {
        let repo = Repository::open(worktree_path)?;

        if create {
            let head = repo.head()?.peel_to_commit()?;
            repo.branch(branch, &head, false)?;
        }

        let obj = repo.revparse_single(&format!("refs/heads/{}", branch))?;
        repo.checkout_tree(&obj, None)?;
        repo.set_head(&format!("refs/heads/{}", branch))?;

        Ok(())
    }

    /// List branches
    pub fn list_branches(path: &str) -> Result<BranchInfo, GitError> {
        let repo = Repository::open(path)?;
        let mut local = Vec::new();
        let mut remote = Vec::new();

        for branch in repo.branches(None)? {
            let (branch, branch_type) = branch?;
            if let Some(name) = branch.name()? {
                match branch_type {
                    BranchType::Local => local.push(name.to_string()),
                    BranchType::Remote => {
                        // Strip "origin/" prefix
                        let stripped = name.strip_prefix("origin/").unwrap_or(name);
                        if stripped != "HEAD" {
                            remote.push(stripped.to_string());
                        }
                    }
                }
            }
        }

        let current = Self::get_current_branch(path)?;

        Ok(BranchInfo {
            local,
            remote,
            current,
        })
    }

    /// Get repository status
    pub fn get_status(path: &str) -> Result<GitStatusInfo, GitError> {
        let repo = Repository::open(path)?;
        let mut opts = StatusOptions::new();
        opts.include_untracked(true);

        let statuses = repo.statuses(Some(&mut opts))?;

        let mut modified = Vec::new();
        let mut staged = Vec::new();
        let mut untracked = Vec::new();

        for entry in statuses.iter() {
            let status = entry.status();
            let file_path = entry.path().unwrap_or_default().to_string();

            if status.is_wt_modified() || status.is_wt_deleted() {
                modified.push(file_path.clone());
            }
            if status.is_index_new() || status.is_index_modified() || status.is_index_deleted() {
                staged.push(file_path.clone());
            }
            if status.is_wt_new() {
                untracked.push(file_path);
            }
        }

        let is_clean = modified.is_empty() && staged.is_empty() && untracked.is_empty();

        // Calculate ahead/behind from upstream
        let (ahead, behind) = Self::get_ahead_behind(&repo).unwrap_or((0, 0));

        Ok(GitStatusInfo {
            is_clean,
            ahead,
            behind,
            modified,
            staged,
            untracked,
        })
    }

    /// Get ahead/behind counts from upstream
    fn get_ahead_behind(repo: &Repository) -> Result<(i32, i32), GitError> {
        let head = repo.head()?;

        if !head.is_branch() {
            return Ok((0, 0));
        }

        let branch_name = head.shorthand().unwrap_or("");
        let branch = repo.find_branch(branch_name, BranchType::Local)?;

        if let Ok(upstream) = branch.upstream() {
            let local_oid = head.target().ok_or_else(|| {
                git2::Error::from_str("No local target")
            })?;

            let upstream_oid = upstream.get().target().ok_or_else(|| {
                git2::Error::from_str("No upstream target")
            })?;

            let (ahead, behind) = repo.graph_ahead_behind(local_oid, upstream_oid)?;
            Ok((ahead as i32, behind as i32))
        } else {
            Ok((0, 0))
        }
    }
}
